pkg/runtime/reconciler.go

```diff
@@ -338,6 +343,43 @@ func (r *resourceReconciler) updateResource(
 	return latest, nil
 }
 
+// lateInitializeResource calls AWSResourceManager.LateInitialize() method and
+// returns the AWSResource with late initialized fields.
+//
+// When the late initialization is delayed for an AWSResource, an error is returned
+// with specific requeue delay to attempt lateInitialization again.
+//
+// This method also adds an annotation to K8s CR, indicating the number of
+// late initialization attempts to correctly calculate exponential backoff delay
+//
+// This method also adds Condition to CR's status indicating status of late initialization.
+func (r *resourceReconciler) lateInitializeResource(
+	ctx context.Context,
+	rm acktypes.AWSResourceManager,
+	latest acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("r.lateInitializeResource")
+	defer exit(err)
+
+	rlog.Enter("rm.LateInitialize")
+	lateInitializedLatest, err := rm.LateInitialize(ctx, latest)
+	rlog.Exit("rm.LateInitialize", err)
+	// Always patch after late initialize because some fields may have been initialized while
+	// others require a retry after some delay.
+	// This patching does not hurt because if there is no diff then 'patchResourceMetadataAndSpec'
+	// acts as a no-op.
+	if ackcompare.IsNotNil(lateInitializedLatest) {
+		patchErr := r.patchResourceMetadataAndSpec(ctx, latest, lateInitializedLatest)
+		// Throw the patching error if reconciler is unable to patch the resource with late initializations
+		if patchErr != nil {
+			err = patchErr
+		}
```

Comment:
This is fine. We'll be shadowing any non-nil `err` that may have been returned on line 367, but that's OK I guess.